src/runner.c

```diff
@@ -94,39 +94,45 @@ void execute(void *data) {
   instance->loop = NULL;
 }
 
-void cleanup(napi_env env, void* data, void* ignored) {
+void napi_cleanup(napi_env env, void* data, void* ignored) {
   fse_t *instance = (fse_t*)data;
   stop(instance);
   CHECK(napi_delete_reference(instance->env, instance->callback) == napi_ok);
   free(instance);
 }
+void async_cleanup(uv_async_t* handle) {
+  free(handle);
+}
 
 void stop(fse_t *instance) {
   if (!instance->loop) return;
   unsigned int count = 0;
   CFRunLoopStop(instance->loop);
   CHECK(!uv_thread_join(&instance->thread));
-  instance->async.data = NULL;
-  uv_close((uv_handle_t *) &instance->async, NULL);
+  instance->async->data = NULL;
   uv_mutex_destroy(&instance->mutex);
+  uv_close((uv_handle_t *) instance->async, async_cleanup);
+  instance->async = NULL;
   CHECK(napi_reference_unref(instance->env, instance->callback, &count) == napi_ok);
 }
 
 napi_value start(napi_env env, const char (*path)[PATH_MAX], napi_value callback) {
   napi_value result;
   fse_t *instance = malloc(sizeof(*instance));
+  instance->async = malloc(sizeof(*instance->async));
```

Comment:
You should probably check that the return value of malloc is not NULL. I'm not sure if malloc() can actually fail on macos or that it just oom-kills the process, but better safe than sorry.